src/pkcs15init/pkcs15-lib.c

```diff
@@ -1664,6 +1664,7 @@ sc_pkcs15init_generate_key(struct sc_pkcs15_card *p15card, struct sc_profile *pr
 
 err:
 	sc_pkcs15_free_object(object);
+	free(pubkey_args.key.alg_id);
```

Comment:
Sorry, I just noticed your above discussion on this. I'll just wait a little while to see if you get to a conclusion...

Reply:
I think in this case it is a not the caller, but the `generate_key()` callback of the pkcs15 emulator.

I did not go checking deeper if and when this is initialized in the drivers, but it looks like there is some different cleaning of this -- it should be easy to get traces with valgrind now (On the side note, it is now not executed in CI now because the regular run already crashes -- we might want to either execute the valgrind in the separate job or even if it would fail -- not sure what would be the best).

I think the it should be investigated slightly deeper rather than implementing fast solution using local variables like proposed initially. On the other hand if it is double free, setting just NULL after the free might solve the issue.

Reply:
The double-free happens because `sc_pkcs15init_store_public_key()` in https://github.com/OpenSC/OpenSC/blob/b031789dccbdc46d151354719f0d527fe100cf76/src/pkcs15init/pkcs15-lib.c#L1654-L1667 releases the `alg_id` via `sc_pkcs15_erase_pubkey()`. Setting `NULL` there should work fine.

Reply:
If I understand correctly, `sc_pkcs15_erase_pubkey()` in `sc_pkcs15init_store_public_key()` is called on _(shallow) copy_,
https://github.com/OpenSC/OpenSC/blob/b031789dccbdc46d151354719f0d527fe100cf76/src/pkcs15init/pkcs15-lib.c#L1858-L1860
so the `0/NULL`s are not propagated outside. The second call to  `sc_pkcs15_erase_pubkey()` then causes another double-free.